### PR TITLE
feat(oracle): add timestamp to exchange rates

### PR DESCRIPTION
**Pull request by gsk967**
## Description

closes: #XXXX

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/oracle/types/asset.go [hunk 1]
```diff
@@ -7,17 +7,17 @@ import (
...
-	USDDenom        string = "USD"
-	BlocksPerMinute        = uint64(10)
-	BlocksPerHour          = BlocksPerMinute * 60
-	BlocksPerDay           = BlocksPerHour * 24
-	BlocksPerWeek          = BlocksPerDay * 7
-	BlocksPerMonth         = BlocksPerDay * 30
-	BlocksPerYear          = BlocksPerDay * 365
-	MicroUnit              = int64(1e6)
+	UmeeDenom       = appparams.BondDenom
+	UmeeSymbol      = "UMEE"
```
Comment (gsk967) [inline x/oracle/types/asset.go:11]: I changed lowercase to UpperCase because we are storing all denoms to Uppercase in our leverage and oracle acceptedlist

File: proto/umee/oracle/v1/oracle.proto [hunk 2]
```diff
@@ -149,8 +149,8 @@ message AvgCounter {
 message ExchangeRate {
   option (gogoproto.equal)            = false;
   option (gogoproto.goproto_stringer) = false;
-
-  string rate = 1 [
+  string denom                        = 1 [(gogoproto.moretags) = "yaml:\"denom\""];
```
[the remark this answers is not in the document]
Comment (gsk967) [reply proto/umee/oracle/v1/oracle.proto:152]: how do we know price for which denom and `ExchangeRate` is single record with timestamp, price and denom

File: x/oracle/abci.go [hunk 3]
```diff
@@ -77,7 +77,8 @@ func CalcPrices(ctx sdk.Context, params types.Params, k keeper.Keeper) error {
 		if err != nil {
 			return err
 		}
-		k.SetExchangeRateWithEvent(ctx, denom, exchangeRate)
+		// save the exchange rate to store with denom and timestamp
+		k.SetExchangeRateWithEvent(ctx, denom, exchangeRate, ctx.BlockTime())
```
[the remark this answers is not in the document]
Comment (gsk967) [reply x/oracle/abci.go:81]: Ok

File: app/test_helpers.go [hunk 4]
```diff
@@ -266,9 +266,11 @@ func IntegrationTestNetworkConfig() network.Config {
 	// execute ballot voting and thus clear out previous exchange rates, since we
 	// are not running a price-feeder.
 	oracleGenState.Params.VotePeriod = 1000
-	oracleGenState.ExchangeRates = append(oracleGenState.ExchangeRates, oracletypes.NewExchangeRateTuple(
```
[the remark this answers is not in the document]
Comment (gsk967) [reply app/test_helpers.go:269]: ok

File: x/oracle/keeper/keeper.go [hunk 5]
```diff
@@ -75,12 +76,12 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
 // denom asset from the store.
 func (k Keeper) GetExchangeRate(ctx sdk.Context, symbol string) (types.ExchangeRate, error) {
-	v := store.GetValue[*types.ExchangeRate](ctx.KVStore(k.storeKey), types.KeyExchangeRate(symbol),
+	v := store.GetValue[*types.DenomExchangeRate](ctx.KVStore(k.storeKey), types.KeyExchangeRate(symbol),
 		"exchange_rate")
 	if v == nil {
 		return types.ExchangeRate{}, types.ErrUnknownDenom.Wrap(symbol)
 	}
-	return *v, nil
+	return types.NewExchangeRate(strings.ToUpper(symbol), sdk.MustNewDecFromStr(v.Rate), v.Timestamp), nil
```
[the remark this answers is not in the document]
Comment (gsk967) [reply x/oracle/keeper/keeper.go:84]: `NewExchangeRate` returns only struct of denom, rate and timestamp 

`*v` is only contains rate and timestamp only

File: x/oracle/types/exchange_rates.go [hunk 6]
```diff
@@ -14,3 +15,25 @@ func NewExchangeRate(denom string, exchangeRate sdk.Dec, t time.Time) ExchangeRa
...
+// DenomExchangeRate is type for storing rate and timestamp of denom into store without denom.
+type DenomExchangeRate struct {
+	Rate      string    `json:"rate"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+// Marshal implements store.Marshalable.
+func (d *DenomExchangeRate) Marshal() ([]byte, error) {
+	return json.Marshal(d)
+}
```
[the remark this answers is not in the document]
Comment (gsk967) [reply x/oracle/types/exchange_rates.go:28]: `json.Marshal` returns binary only I check the get and set and tests everything is working fine

File: x/oracle/types/exchange_rates.go [hunk 7]
```diff
@@ -14,3 +15,25 @@ func NewExchangeRate(denom string, exchangeRate sdk.Dec, t time.Time) ExchangeRa
 		Timestamp: t,
 	}
 }
+
+// DenomExchangeRate is type for storing rate and timestamp of denom into store without denom.
+type DenomExchangeRate struct {
```
[the remark this answers is not in the document]
Comment (gsk967) [reply x/oracle/types/exchange_rates.go:20]: Ok

File: x/oracle/keeper/keeper.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (gsk967) [reply x/oracle/keeper/keeper.go:84]: Ok , I am changing return to `ExchangeRate` only

File: proto/umee/oracle/v1/oracle.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/oracle/v1/oracle.proto:152]: Alternately we can store with denom in state (just a little extra field, so not that big of a problem since it's not per-account storage)

File: x/oracle/keeper/keeper.go [hunk 8]
```diff
@@ -110,39 +111,45 @@ func (k Keeper) GetExchangeRateBase(ctx sdk.Context, denom string) (sdk.Dec, err
...
 	key := types.KeyExchangeRate(denom)
-	val := types.ExchangeRate{Rate: rate, Timestamp: ctx.BlockTime()}
-	err := store.SetValue(ctx.KVStore(k.storeKey), key, &val, "exchange_rate")
+	val := types.ExchangeRate{Rate: rate, Timestamp: t}
+	err := store.SetValue[*types.ExchangeRate](ctx.KVStore(k.storeKey), key, &val, "exchange_rate")
 	util.Panic(err)
 }
 
+// SetExchangeRate sets the consensus exchange rate of USD denominated in the
+// denom asset to the store with timestamp
```
Comment (toteki) [inline x/oracle/keeper/keeper.go:124]:
```suggestion
// denom asset to the store using timestamp from ctx
```

File: x/oracle/keeper/keeper.go [hunk 9]
```diff
@@ -110,39 +111,45 @@ func (k Keeper) GetExchangeRateBase(ctx sdk.Context, denom string) (sdk.Dec, err
 	return exchangeRate.Rate.Quo(powerReduction), nil
 }
 
-// SetExchangeRate sets the consensus exchange rate of USD denominated in the
-// denom asset to the store.
-func (k Keeper) SetExchangeRate(ctx sdk.Context, denom string, rate sdk.Dec) {
+// SetExchangeRateWithTimestamp sets the consensus exchange rate of USD denominated in the
+// denom asset to the store with timestamp but with passing timestamp to func
```
Comment (toteki) [inline x/oracle/keeper/keeper.go:115]:
```suggestion
// denom asset to the store with a timestamp specified instead of using ctx
```

File: x/oracle/keeper/keeper.go [hunk 10]
```diff
@@ -110,39 +111,45 @@ func (k Keeper) GetExchangeRateBase(ctx sdk.Context, denom string) (sdk.Dec, err
...
 	defer iter.Close()
 	prefixLen := len(types.KeyPrefixExchangeRate)
-
 	for ; iter.Valid(); iter.Next() {
 		key := iter.Key()
 		denom := string(key[prefixLen : len(key)-1]) // -1 to remove the null suffix
-		var er types.ExchangeRate
-		k.cdc.MustUnmarshal(iter.Value(), &er)
+		var exgRate types.ExchangeRate
+		err := exgRate.Unmarshal(iter.Value())
```
Comment (toteki) [inline x/oracle/keeper/keeper.go:149]: Any reason for the change from `k.cdc.MustUnmarshal` to `exgRate.Unmarshal`?

File: x/oracle/types/asset.go (hunk 1 above)
Comment (toteki) [reply x/oracle/types/asset.go:11]: Note: [mainnet oracle accept list](https://umee-api.polkachu.com/umee/oracle/v1/params) has some lowercase so be careful with anything that uses those constants

e.g. `"symbol_denom": "umee"` and for `"atom"`, `"stATOM"`, `"stOSMO"` as well.

Mainnet [leverage](https://umee-api.polkachu.com/umee/leverage/v1/registered_tokens) only has (partial) lowercase for `"stATOM"`

File: x/oracle/types/exchange_rates.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (toteki) [reply x/oracle/types/exchange_rates.go:33]: Any reason to not use `k.cdc` codec for this functionality?

**Review by toteki — COMMENTED** [review]
Left comments - my main concerns are codec usage and making sure we don't accidentally mess with upper/lower case in a way that needs a migration or affects mainnet

File: x/oracle/keeper/keeper.go (hunk 10 above)
Comment (gsk967) [reply x/oracle/keeper/keeper.go:149]: `ExchangeRate` is not a proto message , it is just a normal go struct

**Comment by gsk967** [comment]
> Left comments - my main concerns are codec usage and making sure we don't accidentally mess with upper/lower case in a way that needs a migration or affects mainnet

I did not use codec for ExchangeRate because it is only just go struct and I revert the UMEE to umee in defaultParams

File: x/oracle/types/exchange_rates.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+// ExchangeRate is type for storing rate and timestamp of denom into store without denom.
+type ExchangeRate struct {
+	Rate      sdk.Dec   `json:"rate"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+// Marshal implements store.Marshalable.
+func (d *ExchangeRate) Marshal() ([]byte, error) {
+	return json.Marshal(d)
+}
```
Comment (toteki) [inline x/oracle/types/exchange_rates.go:33]: Do we need to worry about any edge cases here, such as `<nil>` sdk.Dec?

Perhaps add test

File: x/oracle/genesis.go [hunk 12]
```diff
@@ -2,6 +2,7 @@ package oracle
 
 import (
 	"fmt"
+	"time"
```
Comment (toteki) [inline x/oracle/genesis.go:6]: Could you add a genesis `export -> import` or `import -> export` test to the module?

I noticed we don't have one here like in the other modules

**Review by toteki — APPROVED** [review]
Check comments - other than that I have no more issues
